application/models/services/QuestionAttributeFetcher.php

```diff
@@ -0,0 +1,158 @@
+<?php
+
+namespace LimeSurvey\Models\Services;
+
+/**
+ * Fetches question attribute definitions from the available providers
+ */
+
+class QuestionAttributeFetcher
+{
+    /** @var \Question the question where the attributes should apply */
+    private $question;
+
+    /** @var array<string,mixed> array of options to pass to the providers */
+    private $options = [];
+
+    /** @var array<QuestionAttributeProvider> array of question attribute providers */
+    private $providers = [];
+
+    public function __construct()
+    {
+        $this->providers = [
+            new CoreQuestionAttributeProvider(),
+            new ThemeQuestionAttributeProvider(),
+            new PluginQuestionAttributeProvider(),
+        ];
+    }
+
+    /**
+     * Returns the question attribute definitions according to the specified filters,
+     * from all available sources.
+     *
+     * @return array<string,array> array of question attribute definitions
+     */
+    public function fetch()
+    {
+        if (empty($this->question)) {
```

Comment:
Valid case? Or exception/assert failure?

Reply:
So you can throw InvalidArgumentException, or, if missing $this->question always is a bug, you can use assert(!empty($this->question), "question attribute should not be empty);

Reply:
Ready for review